nl_preprocessing.py

```diff
@@ -0,0 +1,183 @@
+from __future__ import unicode_literals
+from functools import reduce
+from typing import Any
+
+from extractor import Function
+import re
+import nltk
+
+nltk.download('averaged_perceptron_tagger')
+nltk.download('stopwords')
+nltk.download('wordnet')
+
+# Precompile often used regex
+first_cap_regex = re.compile('(.)([A-Z][a-z]+)')
+all_cap_regex = re.compile('([a-z0-9])([A-Z])')
+
+
+class CleanFunction:
+    def __init__(self, original_function: Function, clean_name: str, clean_docstring: str, clean_arg_names: list,
+                 clean_arg_types: list, clean_return_type: str, clean_return_expr: list):
+        self.original_function = original_function
```

Comment:
Do we really need to keep a reference to the original? This will break the garbage collection and memory will be at a  premium when training LSTMs.

Reply:
Well, it depends, if we want to store the original data then yes, otherwise it's not needed. Just wasn't sure if we should loose data at this point.

I think we will perform the LSTM training on the data from the CSV file, seperately from the data production pipeline, so there shouldn't be any issues with memory or garabage collection (at least not because of this 😉 ).

Reply:
Here's an example of the current output. I'm happy to hear opinions about whether we should keep the original function data or not:

```json
{
        "author": "apache",
        "repo": "incubator-superset",
        "filename": "./projects/apache__incubator-superset/superset/security.py",
        "signature": {
            "original_function": {
                "name": "can_access",
                "docstring": "Return True if the user can access the FAB permission/view, False\notherwise.\n\nNote this method adds protection from has_access failing from missing\npermission/view entries.\n\n:param permission_name: The FAB permission name\n:param view_name: The FAB view-menu name\n:returns: Whether the use can access the FAB permission/view",
                "arg_names": [
                    "self",
                    "permission_name",
                    "view_name"
                ],
                "arg_types": [
                    "",
                    "str",
                    "str"
                ],
                "return_type": "bool",
                "return_expr": [
                    "return self.is_item_public(permission_name, view_name)",
                    "return self._has_view_access(user, permission_name, view_name)"
                ]
            },
            "clean_name": "can access",
            "clean_docstring": "return true user access fab permission view false otherwise. note method add protection access fail miss permission view entries. param permission name fab permission name param view name fab view menu name return whether use access fab permission view",
            "clean_arg_names": [
                "self",
                "permission name",
                "view name"
            ],
            "clean_arg_types": [
                "",
                "str",
                "str"
            ],
            "clean_return_type": "bool",
            "clean_return_expr": [
                "self be item public permission name view name",
                "self. have view access user permission name view name"
            ]
        }
    }
```

Reply:
I would prefer the steps to be separated. The pipeline should imo look like:
file.py --extract--> Function --preprocess--> CleanFunction....

This pipeline is simple and should be a matter of calling map() multiple times in order to get the next result

The "non-clean" fields will not be used, so the function reference can be deleted. If some of the fields should be used, then a clean version of them should be added to CleanFunction.

Reply:
I agree with that pipeline. If we want to have a nonclean version, we can save the output from the extract step.